# Leonidas Avdelas, 03113182 #
## Lab 2, Stochastic Processes ##

## Μέρος 1 ##


In [1]:
import statistics as stat 
## import the library statistics. We will use it to compute the mean and variance of our list
import numpy as np
from numpy.linalg import matrix_power
from simple_markov_chain_lib import markov_chain

In [2]:
p = 1/6

# A dictionary for the initial distibution. 
# We prescribe the initial distribution
init_probs = {1: 1.0} 
 
# A dictionary for the transition probability  matrix. 
# Every state-key corresponds to a list with tuples of (Next_State,Probability) 
markov_table = {
    1: {2: 1/2, 3:1/2},
    2: {1: 1/3, 2: 1/3, 3: 1/3},
    3: {1: p, 2: 2/3-p, 3: 1/3}
}
 
# Ok... we are ready know
# Let's construct a Markov Chain. So let's call the constructor
mc = markov_chain(markov_table, init_probs)

In [3]:
def run_expirament(N, steps):
    ## Experiment parameters
    #N = 200     # number of samples
    #steps = 20   # the target time
    counter = 0  # to count the number of times the event {X_40  = 0} occurs

    ## Simulation
    for i in range(N):
        mc.start()  # new experiment
        for j in range(steps):  mc.move()
        if mc.running_state == 1:  counter += 1

    phat = counter / N

    return phat

In [4]:
results = []
for i in range(50):
    results.append(run_expirament(200, 20))
print('For 200 samples:')
print(
    """ 
    The sample mean is {0:.5f} and the sample variance is {1:.5f}
    """.format(stat.mean(results), stat.variance(results))
)

For 200 samples:
 
    The sample mean is 0.20770 and the sample variance is 0.00095
    


In [5]:
results = []
for i in range(50):
    results.append(run_expirament(20000, 20))
print('For 20k samples:')
print(
    """ 
    The sample mean is {0:.5f} and the sample variance is {1:.5f}
    """.format(stat.mean(results), stat.variance(results))
)

For 20k samples:
 
    The sample mean is 0.20358 and the sample variance is 0.00001
    


In [6]:
init_probs = {3: 1.0}

mc = markov_chain(markov_table, init_probs)

results = []
for i in range(50):
    results.append(run_expirament(20000, 20))
print('For 20k samples:')
print(
    """ 
    The sample mean is {0:.5f} and the sample variance is {1:.5f}
    """.format(stat.mean(results), stat.variance(results))
)

For 20k samples:
 
    The sample mean is 0.20405 and the sample variance is 0.00001
    


# Παραδοτέο 1 #
1. Όπως βλέπουμε και παραπάνω, ο μέσος όρος όταν $N = 200$ είναι $0.213$, ενώ όταν $N = 20000$ είναι $0.203$.
2. Η θεωρητική τιμή που υπολογίζουμε είναι:
$$ π_{20} = π_0 P^{20} = (1, 0, 0)P^{20} $$

Κάνοντας τις πράξεις έχουμε ότι $π_{20} = (0.204, 0.428, 0.367)$. Αρα $π_{20}(1) = 0.204$, τιμή πολύ κοντά σε αυτή που φτάσαμε με την μέθοδο Monte Carlo.
3. H δειγματική διασπορά είναι $0.00083$ όταν $N = 200$, ενώ όταν $N = 20000$ είναι $0.00001$.
4. H εκτίμηση μας δεν αλλάζει, όπως βλέπουμε και από το ακριβώς παραπάνω κελί.

## Μέρος 2 ##

In [7]:
p_a = 0.6
p_b = 1 - p_a
init_probs_2  = {'0-0':1}

P_table = {
    '0-0': {'15-0':p_a, '0-15':p_b},
    '15-0': {'30-0':p_a, '15-15':p_b},
    '0-15': {'15-15': p_a, '0-30':p_b},
    '30-0': {'40-0': p_a, '30-15': p_b},
    '15-15': {'30-15': p_a, '15-30': p_b},
    '0-30': {'15-30': p_a, '0-40': p_b},
    '40-0': {'GameA': p_a, '40-15': p_b},
    '30-15': {'40-15': p_a, 'Deuce': p_b},
    '15-30': {'Deuce': p_a, '15-40': p_b},
    '0-40': {'15-40': p_a, 'GameB': p_b},
    '40-15': {'GameA': p_a, 'AdvA': p_b},
    '15-40': {'AdvB': p_a, 'GameB': p_b},
    'Deuce': {'AdvA': p_a, 'AdvB': p_b},
    'GameA': {'GameA': 1},
    'AdvA': {'GameA': p_a, 'Deuce': p_b},
    'AdvB': {'Deuce': p_a, 'GameB': p_b},
    'GameB': {'GameB': 1}
}

mc2 = markov_chain(P_table, init_probs_2)

In [8]:
## Experiment parameters
N = 10000     # number of samples
steps = 25    # the target time
counter = 0  # to count the number of times the event {X_40  = 0} occurs

## Simulation
for i in range(N):
    mc2.start()  # new experiment
    for j in range(steps):  mc2.move()
    if mc2.running_state == 'GameA':  counter += 1

phat = counter / N

print(
    """
    We executed {0} times the first {1} steps of the markov chain
    and we captured the running state in state GameA {2} times.
    So we estimate the Pr[X_GameA = 1] to be {3}
    """.format(N, steps, counter, phat)
)



    We executed 10000 times the first 25 steps of the markov chain
    and we captured the running state in state GameA 7297 times.
    So we estimate the Pr[X_GameA = 1] to be 0.7297
    


In [9]:
print('End of Lab 2')

End of Lab 2
